Classes/Domain/NodeCreation/NodeCreationService.php

```diff
@@ -38,165 +28,153 @@ class NodeCreationService
      */
     protected $nodeUriPathSegmentGenerator;
 
-    /**
-     * @Flow\Inject
-     * @var PropertyMapper
-     */
-    protected $propertyMapper;
+    private PropertiesHandler $propertiesHandler;
+
+    public function __construct(Context $subgraph, PropertyMapper $propertyMapper)
+    {
+        $this->propertiesHandler = new PropertiesHandler($subgraph, $propertyMapper);
+    }
 
     /**
      * Applies the root template and its descending configured child node templates on the given node.
      * @throws \InvalidArgumentException
      */
-    public function apply(RootTemplate $template, NodeInterface $node, CaughtExceptions $caughtExceptions): void
+    public function createMutatorCollection(RootTemplate $template, ToBeCreatedNode $node, CaughtExceptions $caughtExceptions): NodeMutatorCollection
     {
         $nodeType = $node->getNodeType();
-        $propertiesAndReferences = PropertiesAndReferences::createFromArrayAndTypeDeclarations($this->convertProperties($nodeType, $template->getProperties(), $caughtExceptions), $nodeType);
-
-        // set properties
-        foreach ($propertiesAndReferences->requireValidProperties($nodeType, $caughtExceptions) as $key => $value) {
-            $node->setProperty($key, $value);
-        }
-
-        // set references
-        foreach ($propertiesAndReferences->requireValidReferences($nodeType, $node->getContext(), $caughtExceptions) as $key => $value) {
-            $node->setProperty($key, $value);
-        }
 
-        $this->ensureNodeHasUriPathSegment($node, $template);
-        $this->applyTemplateRecursively($template->getChildNodes(), $node, $caughtExceptions);
+        $properties = $this->propertiesHandler->createdFromArrayByTypeDeclaration($template->getProperties(), $nodeType);
+
+        $validProperties = array_merge(
+            $this->propertiesHandler->requireValidProperties($properties, $caughtExceptions),
```

Comment:
This is hard to read but see my comments in the PropertiesHandler related to that.